Updates to state provider deep dive

trmartin4 opened the pull request:
## 🎟️ Tracking

https://bitwarden.atlassian.net/browse/PM-14197
https://bitwarden.atlassian.net/browse/PM-19220
https://bitwarden.atlassian.net/browse/PM-19218
https://bitwarden.atlassian.net/browse/PM-19219

## 📔 Objective

- Moved the section on `update` out of the "Advanced Usage" section, since we are going to be encouraging teams to use it to avoid unnecessary I/O for state updates (see https://github.com/bitwarden/clients/pull/13271 for an example).  Before we start asking teams to pay more attention to this, I wanted to make sure it had a good place in the docs.
- Re-ordered the definitions to move `ActiveUserState` to the bottom, since it is no longer recommended.
- Removed some of the references to migrations, since they are mostly complete.
- Documented `cleanupDelayMs` option behavior when it's set to `0`.
- Documented that we discourage `firstValueFrom(state$)` directly after update.
- Documented that `state$` does not guarantee an emission on same value updates.

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation
  team

## 🦮 Reviewer guidelines

<!-- Suggested interactions but feel free to use (or not) as you desire! -->

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed
  issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

trmartin4 commented:
❓ @justindbaur Do you think that it's OK to remove the section on Migration and the details on how `StateService` definitions map to `StateProvider` definitions?  I don't want to clutter up the docs with things that teams aren't using today, but I also see the value in having it for reference in the future.  Perhaps a sub-page?  Or are you OK with keeping it in git history?  Or would you prefer to leave it as-is?

justindbaur commented:
@trmartin4 Thank you for this, I had thought about adding some more docs regarding `update` now that we are going to recommend it more.

I do think having a brief blurb about state migrations is good as teams will still occasionally have to do them. We don't need to keep around the state service to state provider mappings though. Since state migrations shouldn't change after being written though it could be useful to just defer to [here](https://github.com/bitwarden/clients/tree/main/libs/common/src/state-migrations/migrations) for any examples and we don't need one here in the docs.

As for `shouldUpdate` I think it might be time for a new example(s). What we are about to ask people is to do more than just avoid saving null when it's already null and instead asking them to avoid saving the same _value_ over and over. So I was thinking about adding something like the following:

=============

A simple example using `shouldUpdate` on simple JavaScript primitive types that work nicely with the strict equal operator (`===`). 

```typescript
const USES_KEYCONNECTOR: UserKeyDefinition<boolean> = ...;

async setUsesKeyConnector(value: boolean, userId: UserId) {
  // Only do the update if the current value saved in state
  // differs in equality of the incoming value.
  await this.stateProvider.getUser(userId, USES_KEYCONNECTOR).update(
    currentValue => currentValue !== value
  );
}
```

Implementing a custom equality operator

```typescript
type Cipher = { id: string, username: string, password: string, revisionDate: Date };
const LAST_USED_CIPHER = UserKeyDefinition<Cipher> = ...;

async setLastUsedCipher(lastUsedCipher: Cipher | null, userId: UserId) {
  await this.stateProvider.getUser(userId, LAST_USED_CIPHER).update(
    currentValue => !this.areEqual(currentValue, lastUsedCipher)
  );
}

areEqual(a: Cipher | null, b: Cipher | null) {
  if (a == null) {
    return b == null;
  }

  if (b == null) {
    return false;
  }

  // Option one - Full equality, comparing every property for value equality
  return a.id === b.id && 
    a.username === b.username && 
    a.password === b.password &&
    a.revisionDate === b.revisionDate;

  // Option two - Partial equality based on requirement that any update would
  // bump the revision date.
  return a.id === b.id && a.revisionDate === b.revisionDate;
}
```

It's important that if you implement an equality function that you then negate the output of that function for use in `shouldUpdate` since you will want to go through the update when they are NOT the same value.

=============

trmartin4 commented:
@justindbaur I addressed your concerns, if you don't mind taking a look.

trmartin4 commented:
@justindbaur added today's feedback 👍

justindbaur commented on docs/architecture/deep-dives/state/index.md:
```diff
@@ -142,212 +143,192 @@ own key definition.
-- [`EnvironmentService`](https://github.com/bitwarden/clients/pull/7621)
-- [Org Keys](https://github.com/bitwarden/clients/pull/7521)
+Use of `firstValueFrom()` should be avoided. If you find yourself trying to use `firstValueFrom()`,
+consider propagating the underlying observable instead of leaving reactivity. :::
```
Missing end to the warning block?

trmartin4 replied:
Oops! Fixed in [`fbd3861` (#549)](https://github.com/bitwarden/contributing-docs/pull/549/commits/fbd3861330cdced948f619c081bea4644cc6ed43).

justindbaur commented on docs/architecture/deep-dives/state/index.md:
```diff
@@ -142,212 +143,192 @@ own key definition.
-where the key will be a user's ID and the value being the legacy account object.
-
-:::
+:::warning `firstValueFrom()` and state updates
```
Somewhere in here we can also mention that `update` will return the new value that was just saved back to the caller, or in the case of a custom `shouldUpdate` function that returned false it will return the existing state value. So if consumers need the updated value they can always get it that way.

trmartin4 replied:
Addressed with [`fbd3861` (#549)](https://github.com/bitwarden/contributing-docs/pull/549/commits/fbd3861330cdced948f619c081bea4644cc6ed43).  Let me know if you think the warning block is too big and should just be a new section.

justindbaur reviewed (commented):
Just a few comments, thank you so much!

trmartin4 commented on docs/architecture/deep-dives/state/index.md:
```diff
@@ -142,212 +143,192 @@ own key definition.
-  async clear(userId: UserId): Promise<void> {
-    await this.stateProvider.getUser(userId, FOLDERS_USER_STATE).update((state) => null);
+  async clearStateValue(userId: UserId): Promise<void> {
+    await this.stateProvider.getUserState$(userId, DOMAIN_USER_STATE).update((state) => null);
```
Fixed in [`fbd3861` (#549)](https://github.com/bitwarden/contributing-docs/pull/549/commits/fbd3861330cdced948f619c081bea4644cc6ed43).  Thank you for catching that.

justindbaur reviewed (approved):
Thank you so much

trmartin4 commented on docs/architecture/deep-dives/state/index.md:
```diff
@@ -142,212 +143,199 @@ own key definition.
-  - A `SingleUserState<T>` method pointing at the same `UserKeyDefinition` as `ActiveUserState` and
-    pointing at the user that is active that had `update` called
-  - Someone updates the key directly on the underlying storage service _(please don't do this)_
+`GlobalState<T>.state$` will emit when the chosen storage location emits an update to the state
```
@MGibson1 I believe I adjusted this in all the proper places, keeping `SingleUserState<T>`, `ActiveUserState<T>`, and `GlobalState<T>` but adding the `T | null` type parameters for the observables.

trmartin4 commented on docs/architecture/deep-dives/state/index.md:
```diff
@@ -142,212 +143,199 @@ own key definition.
+:::warning
 
-type ExpectedAccountState = { myUserData: string };
+`ActiveUserState` has race condition problems. Do not use it for updates and consider transitioning
```
Addressed in daf03ce903fa53c06789778e2ab254c766150d04.

trmartin4 commented on docs/architecture/deep-dives/state/index.md:
```diff
@@ -142,212 +143,199 @@ own key definition.
+If you do need to obtain the result of an update in a non-reactive way, you should use the result
+returned from the `update()` method. This should be used instead of immediately re-requesting the
+value through `firstValueFrom()`. The `update()` will return the value that will be persisted to
+state, after any `shouldUpdate()` filters are applied.
```
Addressed in daf03ce903fa53c06789778e2ab254c766150d04.

trmartin4 commented on docs/architecture/deep-dives/state/index.md:
```diff
@@ -357,16 +345,86 @@ if (accounts?.[userId] == null) {
+- The use of `firstValueFrom` with no `timeout`. Behind the scenes we enforce that the observable
+  given to `combineLatestWith` will emit a value in a timely manner, in this case a `1000ms`
+  timeout, but that number is configurable through the `msTimeout` option.
+- We don't guarantee that your `updateState` function is called the instant that the `update` method
```
Addressed in daf03ce903fa53c06789778e2ab254c766150d04.

trmartin4 commented on docs/architecture/deep-dives/state/index.md:
````diff
@@ -357,16 +345,86 @@ if (accounts?.[userId] == null) {
+  },
+);
+```
+
````
Added this to the section on suggested usages in daf03ce903fa53c06789778e2ab254c766150d04.

trmartin4 commented on docs/architecture/deep-dives/state/index.md:
```diff
@@ -142,212 +143,199 @@ own key definition.
+defined by the corresponding `KeyDefinition`.
 
-### `GlobalState<T>`
+#### `SingleUserState<T>`
```
Could you suggest how you'd like to address `combinedState$`?

trmartin4 commented:
@MGibson1 I have addressed all of your concerns, with the exception of https://github.com/bitwarden/contributing-docs/pull/549#discussion_r2005924611.

I am not sure that we'll actually be merging this PR though, as we may instead move the Deep Dive out of contributing docs and into the markdown documentation for our shiny new state Nx package on `clients` that we're working on carving out.  However, I want to make sure all your concerns are addressed so that I can port over the markdown and be confident that it will be accurate.

trmartin4 commented:
Closing this pull request, as we would like to incorporate the state provider into the markdown documentation for the new `@bitwarden/state` package in `clients`, introduced in https://github.com/bitwarden/clients/pull/15772.
